Ember 4.0

andsmedeiros opened the pull request:
This commit provides support for Ember and EmberData 4.x.

# What was done?

Besides resolving most deprecations and updating dependencies, this is a basis for the modernization of the codebase. 
I attempted to migrate most old JS syntax to modern alternatives while maintaining functionality. Also, old Ember and Data paradigms were mostly shifted out, like EmberObject being ditched where possible. Some quirks and hacks for obsolete Ember/Data versions were removed.

# What was not done?

I left out updating the README or the package version as that should be better handled by the package manager probably.
Also, this does *NOT* attempt to phase out EmberData's adapter paradigm yet.

# Testing

I updated all the testing code also.  However, I cannot get tests to run with scenarios `ember-lts-4.8` and `ember-lts-4.12` due to an error:

```
Global error: Uncaught TypeError: Class extends value undefined is not a constructor
```

This comes off some EmberData's private imports (IIRC `RecordArray` being undefined), but I couldn't spare the time to properly debug it.

Other scenarios run fine.

# Breaking changes

##  TL;DR

Before:
```js
import { Adapter } from 'ember-pouch';
import { PouchDB } from 'ember-pouch/pouchdb';

const db = new PouchDB('my-database');

export default class ApplicationAdapter extends Adapter {
    db = db;
}
```

After:

```js
import { Adapter } from 'ember-pouch';
import { PouchDB } from 'ember-pouch/pouchdb';

const db = new PouchDB('my-database');

export default class ApplicationAdapter extends Adapter {
    constructor(owner) {
        super(owner, db);
    }
}
```

Because `PouchAdapter` relied on `init()` to initialize its change listener, `db` could be set as a class property, as that would be set before `init()` runs. Now, the change listener must be set on the adapter constructor and keeeping `db ` as a class property would mean setting it only after the constructor runs.

The adapter constructor have been now directly defined and takes the database object as parameter. This way, it can set the property itself and correctly set the change listener during construction.

# Afterword

I will also open another PR with a branch aiming Ember and Data >=5 (including 6.0). I am available if anything is needed to getting this working well. 

🖖🏽

andsmedeiros committed:
Fix for Ember v5

andsmedeiros committed:
Moves runtime dependencies (as PouchDB) to from dev dependencies to dependencies

andsmedeiros committed:
Re-add pouchdb-browser as a dependency

andsmedeiros committed:
Removes deprecated Embet.Snapshot#type from PouchAdapter#findHasMany

andsmedeiros committed:
Updating addon to work with Ember >= 6.4.0 and EmberData >= 5.5.0

andsmedeiros committed:
Modernizes syntax, corrects `snapshot.type` deprecation, removes use of private API

andsmedeiros committed:
Removes .idea directory; adds it to .*ignore files

andsmedeiros committed:
Modernizes code

andsmedeiros committed:
Satisfies deprecation ember-data:deprecate-legacy-imports

andsmedeiros committed:
Modernizes syntax

andsmedeiros committed:
Complies to ember-data:deprecate-store-extends-ember-object deprecation

andsmedeiros committed:
Ditches EmberObject and modernizes syntax of `PouchAdapter`.
I attempted to leave the logic untouch as much as possible, converting mostly notation, with a few exceptions:
- The `_onInit` event handler and `init` hooks had to be replaced by logic in the constructor and class fields.
- `_init` method was renamed to `prepare` because it is overridden in test app's adapters and couldn't be made a true private method. Also, naming it `init` would collide with EmberObject hooks.
- Made the constructor take the database object as parameter. Because `init` hooks runs after the constructor, this gave extending adapters enough time to set `this.db`. As we merged those into the constructor,  the change handler would not be set, as `db` would not be defined yet. This is a breaking change.

This commit also updates the test adapters to use this new constructor API.

andsmedeiros committed:
Adds `inverse: null` to `ingredients`

andsmedeiros committed:
Modernizes syntax

andsmedeiros committed:
Modernizes syntax, mostly turning long promise chains into async functions

andsmedeiros committed:
Still working out dependencies

andsmedeiros committed:
Removes instances of `assert.expect()` becase the linter hates it and because it does not prove anything really.

andsmedeiros committed:
QUnit dislikes when tests have no assertions; assert that no exception is thrown.

andsmedeiros committed:
Removes useless `test-support.css` import

andsmedeiros committed:
Prettier pass

andsmedeiros committed:
Corrects nasty error messages in console when deleting indexes in parallel

andsmedeiros committed:
Modernizes syntax and ditches `EmberObject` wrappers in attachments. This is a breaking change.

andsmedeiros committed:
Some cleanup and simplification so the code is clearer and easier to read

andsmedeiros committed:
Refactors flag checking code for more clarity

andsmedeiros committed:
Directly exports the extended `PouchDb` class returned by `.plugin()` calls

andsmedeiros committed:
Avoid defining `ApplicationSerializer` in `/app`; this should probably be re-exportation of `PouchSerializer`.
Instead of providing the application serializer in the addon, defines it only in the dummy app.

andsmedeiros committed:
Small ESLing config corrections

- Corrects ESLint config file name
- dds rule to ignore variables starting with `_`
- Prevents import linting errors in the ESLint config file itself

andsmedeiros committed:
Better consistency of addon exports

- Moves `PouchModel` to `models/pouch.js`
- Exports the model from the new path
- Adds a model export from `/app`
- Exports `PouchDB`, with plugins already attached, directly from the addon

andsmedeiros committed:
Correctly adds decorator configuration in eslint.config.mjs

andsmedeiros committed:
Removes useless dummy app styles

andsmedeiros committed:
Extracts logic from `adapter:taco-salad` into `adapter:application` and makes the first inherit from the latter. Removes some dependencies and modernizes some code.

andsmedeiros committed:
Modernizes dummy app's model classes syntax

andsmedeiros committed:
Removes useless template file

andsmedeiros committed:
Modernizes syntax

andsmedeiros committed:
Removes useless PouchDB export in the dummy ap

andsmedeiros committed:
Removes old ember data feature `ds-references`

andsmedeiros committed:
Modernizes some more syntax and removes useless dependencies. This also attempts to reformat some tests so the intent is clearer.

andsmedeiros committed:
Removes some old dependencies and normalizes some testing behavior.

andsmedeiros committed:
Modernizes some syntax.

andsmedeiros committed:
Ditches EmberObject, as that is not used anymore to create attachments. Added messages to all assertions.

andsmedeiros committed:
This seems it was only used to import PouchDB into broccoli's build pipeline via `app.import`. This is probably not needed anymore since `ember-auto-import` should take care of this.

andsmedeiros committed:
Major [unfinished] overhaul of `PouchAdapter`.
Logic is mostly maintained, but the syntax is being modernized and some behaviour is being refactored to more clearly express intent. Also, some old code has been removed. This includes:
- `#indexPromises` class field
- `indexPromises` argument in `prepare`
- `_schema`, renamed to `schema` simply.

Also, some old quirks to ensure compatibility with ancient Ember and Data verions were removed.
Tests have been also updated to contemplate these modifications.

andsmedeiros committed:
Adds .gitkeep to `dummy/app/styles` so build doesn't break.

andsmedeiros committed:
Some more code cleanup and modernization. Whole adapter is covered now.
Removes useless `extractDeleteRecord()` from util file.

andsmedeiros committed:
Renames `module-for-pouch-acceptance.js` to `install-context.js` to move away from old QUnit APIs. Also, that is makes the file intent more clear.
Updates imports in test files to reflect this.

andsmedeiros committed:
Enables some rules tha had been manually disabled before. Corrected ignoring unused arguments starting with `_`. Bumps JS version to latest.

andsmedeiros committed:
Removes unused CSS and HBS linting scripts

andsmedeiros committed:
Adds scenarios for ember 4.x

andsmedeiros committed:
Removes useless `assets/dummy.css` import

andsmedeiros committed:
Slightly refactors formatting

andsmedeiros committed:
Pins `@ember/test-helpers` to v3

andsmedeiros committed:
Updates `getOwner` import path

andsmedeiros committed:
Moves `unloadedDocumentChanged` from `adapter:application` to `adapter:taco-salad`

andsmedeiros committed:
Removes `StringTransform` as that breaks early v4 compatibility

andsmedeiros committed:
Corrects square bracket access in ArrayProxy returned by relationship

andsmedeiros committed:
Formatting

andsmedeiros committed:
Deletes old  file

andsmedeiros committed:
Updates ember-try config file to only test for ember/data v4.x

andsmedeiros committed:
Embroider won't build if `dummy.css` is absent -- add it back

andsmedeiros committed:
Remove useless test initialization. Also, it seems `loadTests()` was just wrong?

andsmedeiros committed:
Pins relevant versions to ember-v4 compatible dependencies. Removes useless templating dependencies.

backspace commented:
Thanks for all your work on this! I looked over the PR and all seems well, I’m not sure how we should proceed, this library hasn’t seen much activity in a while and I don’t think we have conventions for maintainers to follow.

The lint job failed but based on the error it’s probably a Babel/transpiling thing 🤔

andsmedeiros committed:
Removes Travis config file

andsmedeiros committed:
Updates Github Actions config file
- Uses Node 18 in all jobs
- Swaps yarn for pnpm
- Updates try scenarios with the ones we actually provide

andsmedeiros committed:
Installs pnpm and Chrome for proper test running

andsmedeiros committed:
Aligns implementation to our new API and makes more user friendly

andsmedeiros committed:
Default model blueprint attributes used to provide a consistent model blueprint

andsmedeiros committed:
Deletes useless? file

andsmedeiros commented:
Thanks for the quick response!
I forgot to update the GH actions, so it was installing Node 14 and that doesn't support `??=' and some other syntax.
I also fixed some more blueprint stuff that went by unnoticed.

About the release plans, I think probably we should release two majors, one aiming v4 and other v5+.
Providing a working version to v4 is meaningful for ensuring there is an upgrade path for Ember 3.28 apps (that's actually my case).

broerse commented:
As numerous projects are running Ember 3 and 4 with Ember-Pouch 8.0.0-beta.2, and it is evident that Ember Data 3 is still used in Ember 4 projects, like for instance, LinkedIn uses Ember 4.8.6 and Ember Data 3.28.3 at the moment, indicating that Ember-Pouch should be compatible with both Ember 3 and 4.

I am all for innovation like this, but I am not sure about the breaking changes. Since Ember-Pouch 8.0.0-beta.2 is currently used in Ember 4 projects, would it not be better to skip this innovation for Ember 4 and maintain compatibility with Ember 3? Maybe we can release the current Beta as 8.0.0 and make a breaking 9.0.0-beta.1 Ember 5+ with the Ember 5.0 #268 PR?

andsmedeiros commented:
@broerse TBH I was under the impression that 8.0.0-beta.2 was not working as intended. Anyway, if it works, I don't see a problem in releasing that as major and skipping this PR, going straight to https://github.com/pouchdb-community/ember-pouch/pull/268 as the next major.

broerse commented:
@andsmedeiros See: https://github.com/broerse/ember-cli-blog . If we release 9.0.0-beta.1 for Ember 5+ It would be super if you have time to do a PR on ember-cli-blog to bring it to the newest Ember just to demo it.
